Methods that remove batch effects while retaining group differences may lead to exaggerated confidence in downstream analyses
========================================================


```{r echo=FALSE, results='hide'}
"
Alternativer til tittel
Methods that remove batch effects while retaining group differences may lead to incorrect conclusions in downstream analyses
Methods that remove batch effects while retaining group differences may lead to incorrect conclusions
"
```

## Abstract
Removal of, or adjustment for, batch effects or centre differences is generally required when such effects are present in data. In particular, when preparing microarray gene expression data from multiple cohorts, array platforms, or batches for later analyses, batch effects can have confounding effects. Many methods and tools exist for this purpose. One method, ComBat which is part of the R package sva, is particularly popular due to its ability to remove batch differences even when batches are small and heterogeneous, while retaining the difference between study groups, e.g. treatments. Unfortunately, this recommended and frequently used approach to batch correction will systematically induce incorrect group differences in downstream analyses when groups are unevenly distributed between the batches. The scientific community seems to be largely unaware of this problem, which most likely has contributed to false discoveries being presented in the published literature.

## Introduction

Extraneous variables, if left unaccounted for, have the potential to lead an investigator into drawing wrong conclusions. In molecular biology, extraneous variables are often called "batch effects", probably due to the fact that reagents and other equipment, for instance microarray chips, are made in batches, and this is frequently observed as an effect in the measurements. See 20676067 for more examples. 

For a typical experiment comparing group differences, the presence of batch effects will decrease the statistical power. If the batch/group design is unbalanced - i.e the proportion of one groups samples belonging to each batch is not equal for all groups - batch may act as a confounder and lead to false positive results as well (20838408).

The standard way to handle an extraneous variable is to include it in the statistical model employed in the inquiry. However, many analytic tools for high throughput data does not cater for this option, and when available it could still be outside the competence of the investigator. Therefore, an alternative two step procedure has emerged, first the batch effects are estimated and removed, creating a "batch effect free" data set. In the next step, the statistical analysis is performed on the adjusted data as there never were batch effects in the first place. This appealing compartmentalization is also convenient for practical purposes, for example when data-processing and statistical analyses are performed by different personnel.


The first step can be achieved by subtracting the mean of the measurements in one batch from all measurements in that batch, i.e mean adjustment or one-way ANOVA adjustment as implemented in the method pamr.batchadjust from the pamr package in R. When the batch/group design is balanced, mean-adjustment will take out some, but not necessarily all, variance attributed to batch and leave the between group variance, thus increasing the statistical power. However, when batch is a confounder, some of the variance removed will be group variance and statistical power will be lost. In addition if there are many groups, some effect will be transferred between groups that share batches, leading to false positive results (_Dette problemet er ikke beskrevet i de andre artiklene eller i einars mail, men saa vidt jeg skjoenner saa kan dette skje_). 

To mitigate the above problems, group differences can, simplified explained, be estimated based on the balanced part of the data and applied to all samples in the adjusted data set. If all the samples, and not only the balanced ones, are used as sample size in later a group comparison, the confidence will be overestimated.

The ComBat method described in 16632515 and also included in the sva package (22257669) implements the above described procedure in an attempt to retain group differences when adjusting for batch effects. Based on actual use of ComBat by the authors and others, we suspect the adjusted data predominantly is treated as "batch effect free" in the subsequent analysis. And as a consequence, confidence is overestimated and false results are reported.



```{r echo=FALSE, results='hide'}
"
Kort oppsummere behovet for batch-korrigering: generelt og i microarraydata.

Kort beskrive metoder som gjoer dette (batch-sentrering, ANOVA, ComBat).

Forklare tilsiktet effekt disse paa batch+gruppe-forskjeller (balansert versus ikke-balansert).

Forklare kort hva faktisk effekt er: ANOVA gir redusert gruppe-forskjeller, mens for ComBat overestimeres den statistiske styrken/sikkerheten av forskjellen.
"
```

## A simple sanity check
The undesired consequences of preserving group effects when correcting for batch effect is readily illustrated with a sanity check using random numbers . The documentation accompanying the sva library has a runnable example demonstrating how to adjust a data set with ComBat followed by a F-test. Swapping the real data with random numbers from a normal distribution (mean=0, sd=1), but otherwise following the instructions, will generate the p-value distribution shown in Figure 1. The skewed distribution is for us a clear indication that this approach may have a unintentional adverse effect. _Jeg har ikke beskrevet permutatsjon checken fordi jeg er usikker paa om vi skal bruke den._

![fig. 1](../reanalysis/leek_sva/figure/pvaluesleeksva.svg)  
**Figure 1**  
Two sanity checks were the recommended use of ComBat fails. Adapted from the user guide in the sva package. 
- a) Real data is substituted with random numbers from a normal distribution (mean=0, sd=1), but the batch/group design is retained, followed by batch adjustment in ComBat and a F-test.
- b) 10 runs of real data with the "cancer" labels permuted within batches, followed by ComBat adjustment and a F-test. All permutations produces a skewed p-value distribution.


## Formler og ord
_Forklare hva som gaar galt med formler og ord. Overlater til Einar aa skrive dette_

## Examples of undesired consequences
As the amount of false positive results when trying to retain group differences depends on the group/batch balance, we will show two examples with varying degree of unbalancedness. 

1) In the first experiment(24421904), cells were treated with glatiramer acetate (a medicine for multiple sclerosis) or a generic and mRNA was measured using microarrays alongside control samples. A batch effect correlating to the chip (Illumina WG-6_V2, six samples per chip) was observed and adjusted for with ComBat, whereafter the data was tested for differetially expressed genes, yielding a list of 1000 genes (Table S5 24421904). Unfortunately the batch/treatment design was unbalanced with several batches having only one of the main treatments of interest. When we re-analyzed their data without using ComBat, but instead blocked for batch effect in limma, only 9 genes were found (FDR<0.05). Additional sanity checks with random numbers or perumted labels were also carried out and the distribution of p-values for differents settings are shown in Fig. 2.  Our conclusion is that most of the genes reported as differtially expresssed in (24421904) are false positives. This example is a sort of "worst case" scenario for applying ComBat, since it both has a very unbalanced batch/group design and a a priori assumpotion of no difference. The R-code for our analysis and a more extensive report can be downloaded from github (https://github.com/vegardny/combat_tests.git).

![fig. 2](../reanalysis/Fadi_Towfic_generic_medicine/figure/pvaluestowfic.svg)
**Figure 2**.
P-value distribution for the main comparison in (data set 2 24421904) glatiramer acetate vs. generic.  
- a) Real data adjusted by ComBat as described in the article followed by a significance test in limma. 
- b) Real data not adjusted by ComBat, tested for significance in limma using batch as a blocking factor. 
- c) Real data is substituted with random numbers from a normal distribution (mean=0, sd=1), but the batch/group design is retained, followed by batch adjustment in ComBat and a significance test in limma.
- d) 10 runs of real data with the glatiramer acetate("DP") and generic("N") labels permuted within batches, followed by ComBat adjustment and significance test in limma.  
_jeg tok med begge sanitycheckene her. Er d) slik einar foreslo?. Merk jeg har ikke permutert alle labeler, bare de mellom DP og N. Det er en del andre._

2) The second example is taken from the supporting information for the original ComBat article (16632515) where it is denoted "data set 2". Cells inhibited for the expression of the TAL1 gene were compared to controls on a microarray platform. The experiment was conducted on three different time points (used as batches) with a total of 30 samples and a fairly balanced batch/treatment set up (6-2, 3-4 and 9-6). ComBat was applied followed by a t-test in order to identify differentially expressed genes. First, we reproduced their analysis including the adjustment by ComBat, but using limma instead of the t-test, resulting in 1003 probes (q<0.05).  Then, we analysed their data without batch adjustment in ComBat, but blocking for batch in limma, resulting in 377 probes (q<0.05). In addition the two sanity checks outline above were performed. In contrast to the above results obtained for(24421904), the P-value distributions for the alternative analyses (Fig.3 a,b) does not indicate a huge difference. Nevertheless, we belive that  P-values are deflated for the ComBat adjusted analysis.

![fig. 3](../reanalysis/ComBat_original/figure/pvaluesjohnson.svg)  
**Figure 3**.
P-value distribution for the main comparison in (data set 2 16632515) for  
- a) Real data adjusted by ComBat as described in the article followed by a significance test in limma. 
- b) Real data not adjusted by ComBat, tested for significance in limma using batch as a blocking factor. 
- c) Real data is substituted with random numbers from a normal distribution (mean=0, sd=1), but the batch/group design is retained, followed by batch adjustment in ComBat and a significance test in limma.
- d) 10 runs of real data with the group labels permuted within batches, followed by ComBat adjustment and significance test in limma.   
_problemet med permutatsjons sanitychechen slik jeg har utfoert og plottet den her er at den neppe kan sies aa feile. Dessuten kan de 3 foerst plottene bli plottet i ett som vist under i en alternativ version jeg synes er finere, men dette gaar ikke for permutasjonstesten tror jeg_

![fig. 4](../reanalysis/ComBat_original/figure/pvaluesjohnson2.svg)   
_Alternativ illustrasjon av de tre p-value distribusjonene_


3) _Her kan vi fylle paa med flere eksempler etter samme mal hvis det trengs. Feks en fersk artikkel fra Nature Genetic([24584070](http://www.nature.com/ng/journal/vaop/ncurrent/full/ng.2916.html)). Der har de kombinert sekvenseringsdata  (deres data) fra 9 lymfom pasienter med et (eksternt) microarray data set som har 6 lymfom proever og 5 kontroller, dvs kontroll gruppen er helt fravaerende hos dem. Dvs veldig ubalansert (9/0, 6/5), MEN skille mellom phenotypene er saa stor at de sikkert hadde funnet omtrent de samme genene signifikante med aa bare bruke de balanserte dataene (men det var jo ikke deres.) Dette er ikke en viktig del av paperet, men det er veldig feil selvom konsekvensense ikke blir store. Dog gir det et falskt inntrykk av at dataene deres ligner tidligere data, se heatmap i suppl.. Dessuten lager de GO analyser av dette som egentlig er 7 aar gamle andres data og ikke deres_

## Discussion

## Increased emphasis on preserving group difference
In the original ComBat article(2007) it is clear that the primary motivation behind ComBat was to employ an Emprical Bayes method for batch-effect removal. The feature of retaining group differences for unbalanced designs is optional and seems to be subordinate, only exemplified in the supplementary information. However, over the years this feature became more important judging from advice given by the author to users
(
[1](https://groups.google.com/d/msg/combat-user-forum/eSVSKwGtuyE/ZIWV2juYmmAJ)
,[2](https://groups.google.com/d/msg/combat-user-forum/Vkb9p7wekd4/h5Etie7FTVQJ)
,[3](https://groups.google.com/d/msg/combat-user-forum/fpBTcgDjiR8/uo4QIZL4sZgJ)
,[4](https://groups.google.com/d/msg/combat-user-forum/26FZlgU2LFQ/W6U_Lhh_64EJ)
). _Aa linke direkte er kanskje litt ufint, alternativt kan det linkes til forumet uten konkrete innlegg_.  
And when ComBat was incorporated in the sva package ( [22257669](http://bioinformatics.oxfordjournals.org/cgi/pmidlookup?view=long&pmid=22257669), 2012), inclusion of group labels was made almost mandatory (an undocumented option of passing a NULL value exists). Thus, this problematic use has likely been common.


## Motivation for this warning
Our knowledge of ComBat came through a typical use case when trying to salvage unbalanced data which had batch effects. Upon realizing that the confidence on our group differences were exaggerated, the literature was searched for a better understanding of correct use and potential overseen limitations of ComBat. But the authors of ComBat and the sva package recommended our usage (16632515, 22257669). In addition, other works looking into the problem of batch effects were mostly recommending ComBat without much concern (22682473, 22133085). A brief inquiry into  some of the articles citing ComBat (574 Google Scholar) revealed few problems, and their method descriptions regarding ComBat were mostly sparse, limited to one or two sentences (24391845, 18414638, 21731603,23630272, 24584070). A further indication of their carefree use of this potentially devastating procedure was the frequent neglect to state the program parameters, i.e batch labels (18414638, 21731603) or group labels (24391845, 18414638, 21731603, 23630272, 23482648, 24584070). Often no effort was done in order to substantiate the existence of batch effects in their data, except for stating the presence of batches (24391845?, 18414638, 21731603, 23630272). In one instance ComBat was even applied on data were effects due to batch were investigated, but not found (_husker ikke ref_). The incorporation of the method into analysis pipelines (16642009,TCGA) and other packages (23452776, 21937664) could make its usage even more trivial and parameters setting harder to perceive.
Taken together we fear that many published results from data adjusted by ComBat are completely or partially false. And knowing that scientists don't give up where one analysis fails, a method that almost ensures a result given a sufficiently unbalanced design will continue to be used.


## Practical advice
We have shown that adjusting for batch effects while preserve the group difference may lead to varying degree of false results. Knowing this, to what degree can an investigator trust a result from a work applying such a method? Essentially, when the batch/group configuration is balanced, or group difference is ignored  (i.e. no group labels not given as parameters to ComBat), problems related to preserving group differences will not occur. For other cases, a re-analysis without using this approach is the most rigorous path. However, this thoroughness is not feasible if the downstream  analysis can not adjust for batch effects by it self. To reach a reliable result, batch effects need to be handled in some way or another. To make matters worse, a re-analysis relies on the availability of the raw data and a description of processing and analysis steps taken in the original work. Even when this is available, the necessary bioinformatic skills and work hours could still be in short supply. For such situations, a superficial assessment can be performed, taking special note of batches were groups of interest are near missing and how likely a group difference is. In essence asking if the balanced parts (effective sample size?) of the data has enough power to detect the presumed effects and if this is the case (16632515, 24584070), treat the results more like an ordered list with the most likely true positives on top while de-emphasizing the somewhat deflated p-values . In contrast, if biological knowledge suggest that a group effect is unlikely  (24391845, 18414638?, 21731603, 23630272), an intermediate lack of batch/group balance  could lead to a mostly false result.

## Reproducible research
_vi boer klare aa tilfredstille kravene i http://biostatistics.oxfordjournals.org/content/10/3/405.full_

## References

http://biostatistics.oxfordjournals.org/content/8/1/118.abstract
Adjusting batch effects in microarray expression data using empirical Bayes methods.
Johnson WE, Li C, Rabinovic A.
Biostatistics. 2007 Jan;8(1):118-27. Epub 2006 Apr 21.
PMID: 16632515 [PubMed - indexed for MEDLINE] Free Article

Molecular profiling of single Sca-1+/CD34+,- cells--the putative murine lung stem cells.
Hittinger M, Czyz ZT, Huesemann Y, Maneck M, Botteron C, Kaeufl S, Klein CA, Polzer B.
PLoS One. 2013 Dec 31;8(12):e83917. doi: 10.1371/journal.pone.0083917. eCollection 2013.
PMID: 24391845 [PubMed - in process] 
http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0083917

Drinking-water arsenic exposure modulates gene expression in human lymphocytes from a U.S. population.
Andrew AS, Jewell DA, Mason RA, Whitfield ML, Moore JH, Karagas MR.
Environ Health Perspect. 2008 Apr;116(4):524-31. doi: 10.1289/ehp.10861.
PMID: 18414638 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2290973/


Epigenetic predictor of age.
Bocklandt S, Lin W, Sehl ME, Sánchez FJ, Sinsheimer JS, Horvath S, Vilain E.
PLoS One. 2011;6(6):e14821. doi: 10.1371/journal.pone.0014821. Epub 2011 Jun 22.
PMID: 21731603 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0014821


Childhood maltreatment is associated with distinct genomic and epigenetic profiles in posttraumatic stress disorder.
Mehta D, Klengel T, Conneely KN, Smith AK, Altmann A, Pace TW, Rex-Haffner M, Loeschner A, Gonik M, Mercer KB, Bradley B, Müller-Myhsok B, Ressler KJ, Binder EB.
Proc Natl Acad Sci U S A. 2013 May 14;110(20):8302-7. doi: 10.1073/pnas.1217750110. Epub 2013 Apr 29.
PMID: 23630272 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.pnas.org/content/110/20/8302.full


Patterns of regulatory activity across diverse human cell types predict tissue identity, transcription factor binding, and long-range interactions.
Sheffield NC, Thurman RE, Song L, Safi A, Stamatoyannopoulos JA, Lenhard B, Crawford GE, Furey TS.
Genome Res. 2013 May;23(5):777-88. doi: 10.1101/gr.152140.112. Epub 2013 Mar 12.
PMID: 23482648 [PubMed - indexed for MEDLINE] Free PMC Article
http://genome.cshlp.org/content/23/5/777.full



A recurrent inactivating mutation in RHOA GTPase in angioimmunoblastic T cell lymphoma.
Yoo HY, Sung MK, Lee SH, Kim S, Lee H, Park S, Kim SC, Lee B, Rho K, Lee JE, Cho KH, Kim W, Ju H, Kim J, Kim SJ, Kim WS, Lee S, Ko YH.
Nat Genet. 2014 Mar 2. doi: 10.1038/ng.2916. [Epub ahead of print]
PMID: 24584070 [PubMed - as supplied by publisher]
http://www.nature.com/ng/journal/vaop/ncurrent/full/ng.2916.html

A recurrent inactivating mutation in RHOA GTPase in angioimmunoblastic T cell lymphoma.
Yoo HY, Sung MK, Lee SH, Kim S, Lee H, Park S, Kim SC, Lee B, Rho K, Lee JE, Cho KH, Kim W, Ju H, Kim J, Kim SJ, Kim WS, Lee S, Ko YH.
Nat Genet. 2014 Mar 2. doi: 10.1038/ng.2916. [Epub ahead of print]
PMID: 24584070 [PubMed - as supplied by publisher]



Batch correction of microarray data substantially improves the identification of genes differentially expressed in rheumatoid arthritis and osteoarthritis.
Kupfer P, Guthke R, Pohlers D, Huber R, Koczan D, Kinne RW.
BMC Med Genomics. 2012 Jun 8;5:23. doi: 10.1186/1755-8794-5-23.
PMID: 22682473 [PubMed - indexed for MEDLINE] Free PMC Article
Related citations
http://www.biomedcentral.com/1755-8794/5/23


Relative impact of key sources of systematic noise in Affymetrix and Illumina gene-expression microarray experiments.
Kitchen RR, Sabine VS, Simen AA, Dixon JM, Bartlett JM, Sims AH.
BMC Genomics. 2011 Dec 1;12:589. doi: 10.1186/1471-2164-12-589.
PMID: 22133085 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.biomedcentral.com/1471-2164/12/589/

Tackling the widespread and critical impact of batch effects in high-throughput data.
Leek JT, Scharpf RB, Bravo HC, Simcha D, Langmead B, Johnson WE, Geman D, Baggerly K, Irizarry RA.
Nat Rev Genet. 2010 Oct;11(10):733-9. doi: 10.1038/nrg2825. Epub 2010 Sep 14.
PMID: 20838408 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.nature.com/nrg/journal/v11/n10/abs/nrg2825.html

A comparison of batch effect removal methods for enhancement of prediction performance using MAQC-II microarray gene expression data.
Luo J, Schumacher M, Scherer A, Sanoudou D, Megherbi D, Davison T, Shi T, Tong W, Shi L, Hong H, Zhao C, Elloumi F, Shi W, Thomas R, Lin S, Tillinghast G, Liu G, Zhou Y, Herman D, Li Y, Deng Y, Fang H, Bushel P, Woods M, Zhang J.
Pharmacogenomics J. 2010 Aug;10(4):278-91. doi: 10.1038/tpj.2010.57.
PMID: 20676067 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.nature.com/tpj/journal/v10/n4/full/tpj201057a.html

Removing batch effects in analysis of expression microarray data: an evaluation of six batch adjustment methods.
Chen C, Grennan K, Badner J, Zhang D, Gershon E, Jin L, Liu C.
PLoS One. 2011 Feb 28;6(2):e17238. doi: 10.1371/journal.pone.0017238.
PMID: 21386892 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0017238

T. Hastie, R. Tibshirani, Balasubramanian Narasimhan and Gil Chu (2013). pamr: Pam: prediction analysis for microarrays. R package version
  1.54.1. http://CRAN.R-project.org/package=pamr
  

Expanding the understanding of biases in development of clinical-grade molecular signatures: a case study in acute respiratory viral infections.
Lytkin NI, McVoy L, Weitkamp JH, Aliferis CF, Statnikov A.
PLoS One. 2011;6(6):e20662. doi: 10.1371/journal.pone.0020662. Epub 2011 Jun 1.
PMID: 21673802 [PubMed - indexed for MEDLINE] Free PMC Article
http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0020662


TCGA Batch Effects Tool
http://bioinformatics.mdanderson.org/main/TCGABatchEffects:Overview

GenePattern 2.0.
Reich M, Liefeld T, Gould J, Lerner J, Tamayo P, Mesirov JP.
Nat Genet. 2006 May;38(5):500-1. No abstract available.
PMID: 16642009 [PubMed - indexed for MEDLINE]
http://www.nature.com/ng/journal/v38/n5/full/ng0506-500.html


The sva package for removing batch effects and other unwanted variation in high-throughput experiments.
Leek JT, Johnson WE, Parker HS, Jaffe AE, Storey JD.
Bioinformatics. 2012 Mar 15;28(6):882-3. doi: 10.1093/bioinformatics/bts034. Epub 2012 Jan 17.
PMID: 22257669 [PubMed - indexed for MEDLINE] Free PMC Article
http://bioinformatics.oxfordjournals.org/cgi/pmidlookup?view=long&pmid=22257669

virtualArray: a R/bioconductor package to merge raw data from different microarray platforms.
Heider A, Alt R.
BMC Bioinformatics. 2013 Mar 2;14:75. doi: 10.1186/1471-2105-14-75.
PMID: 23452776 [PubMed - indexed for MEDLINE] Free PMC Article
http://bioconductor.org/packages/release/bioc/html/virtualArray.html


inSilicoDb: an R/Bioconductor package for accessing human Affymetrix expert-curated datasets from GEO.
Taminau J, Steenhoff D, Coletta A, Meganck S, Lazar C, de Schaetzen V, Duque R, Molter C, Bersini H, Nowé A, Weiss Solís DY.
Bioinformatics. 2011 Nov 15;27(22):3204-5. doi: 10.1093/bioinformatics/btr529. Epub 2011 Sep 21.
PMID: 21937664 [PubMed - indexed for MEDLINE] Free Article


Comparing the biological impact of glatiramer acetate with the biological impact of a generic.
Towfic F, Funt JM, Fowler KD, Bakshi S, Blaugrund E, Artyomov MN, Hayden MR, Ladkani D, Schwartz R, Zeskind B.
PLoS One. 2014 Jan 8;9(1):e83757. doi: 10.1371/journal.pone.0083757. eCollection 2014.
PMID: 24421904 [PubMed - in process] Free PMC Article
http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0083757



